src/facade/reply_builder.h

```diff
@@ -167,54 +167,119 @@ class SinkReplyBuilder {
+    SinkReplyBuilder2* rb;
+  };
+
+ public:
+  constexpr static size_t kMaxInlineSize = 32;
+  constexpr static size_t kMaxBufferSize = 8192;
+
   explicit SinkReplyBuilder2(io::Sink* sink) : sink_(sink) {
   }
 
   // Use with care: All send calls within a scope must keep their data alive!
   // This allows to fully eliminate copies for batches of data by using vectorized io.
-  struct ReplyScope {
-    explicit ReplyScope(SinkReplyBuilder2* rb) : prev_scoped(rb->scoped_), rb(rb) {
-      rb->scoped_ = true;
+  struct ReplyScope : GuardBase {
+    explicit ReplyScope(SinkReplyBuilder2* rb) : GuardBase{std::exchange(rb->scoped_, true), rb} {
     }
-    ~ReplyScope() {
-      if (!prev_scoped) {
-        rb->scoped_ = false;
-        rb->FinishScope();
-      }
+
+    ~ReplyScope();
+  };
+
+  // Reduce number of send calls by aggregating responses in a buffer. Prefer ReplyScope
+  // if it's conditions are met.
+  struct ReplyAggregator : GuardBase {
+    explicit ReplyAggregator(SinkReplyBuilder2* rb)
+        : GuardBase{std::exchange(rb->batched_, true), rb} {
     }
 
-   private:
-    bool prev_scoped;
-    SinkReplyBuilder2* rb;
+    ~ReplyAggregator();
   };
 
- public:
-  void Write(std::string_view str);
+  void Flush();  // Send all accumulated data and reset to clear state
+
+  std::error_code GetError() const {
+    return ec_;
+  }
+
+  size_t UsedMemory() const {
+    return buffer_.Capacity();
+  }
+
+  bool IsSendActive() {
+    return send_active_;
+  }
+
+  void SetBatchMode(bool b) {
+    batched_ = b;
+  }
+
+  void CloseConnection();
+
+  static const ReplyStats& GetThreadLocalStats() {
+    return tl_facade_stats->reply_stats;
+  }
+
+  static void ResetThreadLocalStats() {
+    tl_facade_stats->reply_stats = {};
+  }
+
+ public:  // High level interface
+  virtual void SendLong(long val) = 0;
+  virtual void SendSimpleString(std::string_view str) = 0;
+
+  virtual void SendStored() = 0;
+  virtual void SendSetSkipped() = 0;
+  void SendOk() {
+    SendSimpleString("OK");
+  }
+
+  virtual void SendError(std::string_view str, std::string_view type = {}) = 0;  // MC and Redis
+  void SendError(OpStatus status);
+  void SendError(ErrorReply error);
+  virtual void SendProtocolError(std::string_view str) = 0;
 
  protected:
-  void Flush();        // Send all accumulated data and reset to clear state
+  // Write strings beyond the inline size as refs. Once the scope ends, they're flushed
+  // or converted into pieces if batching is enabled and the buffer is not filled up enough.
+  void WriteI(std::string_view str) {
+    str.size() > kMaxInlineSize ? WriteRef(str) : WritePiece(str);
+  }
+
+  // Constexpr arrays are assumed to be protocol control sequences, stash them as pieces
+  template <size_t S> void WriteI(const char (&arr)[S]) {
+    WritePiece(std::string_view{arr, S - 1});  // we assume null termination
+  }
+
+  template <typename... Args> void Write(Args&&... strs) {
+    (WriteI(strs), ...);
+  }
```

Comment:
Getting rid of branches and loops for Write(...) adds ~10% for sending lots of small data where the cost of copying control sequences is high